A36-xds-for-servers.md

```diff
@@ -0,0 +1,332 @@
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+Bring xDS-based configuration to Servers. xDS has many features and this gRFC
+will not provide support for any of them in particular. However, it will provide
+the central plumbing and APIs to add support for specific features and have them
+work without additional user code changes.
+
+There's multiple pieces involved:
+
+*  The API surface exposed to users to enable the new system.
+*  The gRPC-internal plumbing to allow injecting behavior into servers, and
+   changing that behavior on-the-fly.
+*  The specific xDS protocol behavior gRPC will use to retrieve configuration.
+
+## Background
+
+Since [gRFC A27: xDS-Based Global Load Balancing][A27], Channels have had
+growing support for understanding xDS configuration and configuring themselves.
+However, Servers are also important to connect to xDS in order to power TLS
+communication, authorization, fault injection, rate limiting, and other
+miscellaneous features.
+
+On client-side, users are able to opt-in to the behavior via the `xds:` scheme
+which enables the xDS Name Resolver. Servers in gRPC do not use Name Resolution
+and certainly don't have things like Load Balancing APIs which have been
+leveraged heavily on the client-side design. While server-side will need to use
+a different approach, users should have a similar experience: they make a
+trivial code change to opt-in to xDS, and from that point on newly-added
+xDS-powered gRPC features can be used without code changes.
+
+### Related Proposals:
+
+ * [A27: xDS-Based Global Load Balancing][A27]
+ * [A29: xDS-Based Security for gRPC Clients and Servers][A29]
+
+[A27]: A27-xds-global-load-balancing.md
+[A29]: A29-xds-tls-security.md
+
+## Proposal
+
+Each language will create an "XdsServer" API that will wrap the normal "Server"
+API. This will allow it to see the listening ports, start event, stop event, and
+also allow it access to the server to change behavior (e.g., via an
+interceptor). Users will use the XdsServer API to construct the server instead
+of the existing API.
+
+Credential configuration will be managed separately by providing
+XdsServerCredentials, similar to client-side. The user must pass the
+XdsServerCredentials to the XdsServer API to enable control-plane managed TLS
+certificates. While the user-facing API is discussed here, the precise
+implementation details are covered in gRFC A29. The user is free to pass other
+ServerCredentials types to XdsServer and they will be honored. Note that RBAC is
+authz, not authn, so it would be enabled by using the XdsServer, even without
+XdsServerCredentials.
+
+When starting the XdsServer, it will start communication with xDS to receive
+initial configuration. `bind()` and `listen()` may happen immediately, but
+`accept()`ing on the port must be delayed until the initial configuration is
+received. `listen()` should be delayed if possible, but commonly needs to happen
```

Comment:
Do we still want to delay `Accept` before the initial configuration is received, or do the `Accept`+`Close` approach that we decided if we receive a bad config?  cc @markdroth

Reply:
See my comment from yesterday.  I suggested a fairly long explanation of all of the constraints around when to call `bind()`, `listen()`, `accept()`, etc.